File: backend/src/main/kotlin/se/daan/moneyprinters/model/game/api/Event.kt
package se.daan.moneyprinters.model.game.api

import com.fasterxml.jackson.annotation.JsonSubTypes
import com.fasterxml.jackson.annotation.JsonTypeInfo

@JsonTypeInfo(
        use = JsonTypeInfo.Id.NAME,
        include = JsonTypeInfo.As.PROPERTY,
        property = "type"
)
@JsonSubTypes(
        value = [
            JsonSubTypes.Type(value = GameCreated::class, name = "GameCreated"),
            JsonSubTypes.Type(value = PlayerAdded::class, name = "PlayerAdded"),
            JsonSubTypes.Type(value = PromotedToGameMaster::class, name = "PromotedToGameMaster"),
            JsonSubTypes.Type(value = GameStarted::class, name = "GameStarted"),
            JsonSubTypes.Type(value = NewTurnStarted::class, name = "NewTurnStarted"),
            JsonSubTypes.Type(value = DiceRolled::class, name = "DiceRolled"),
            JsonSubTypes.Type(value = StartMoneyReceived::class, name = "StartMoneyReceived"),
            JsonSubTypes.Type(value = LandedOnSafeSpace::class, name = "LandedOnSafeSpace"),
            JsonSubTypes.Type(value = LandedOnBuyableSpace::class, name = "LandedOnBuyableSpace"),
            JsonSubTypes.Type(value = SpaceBought::class, name = "SpaceBought"),
            JsonSubTypes.Type(value = BidStarted::class, name = "BidStarted"),
            JsonSubTypes.Type(value = BidPlaced::class, name = "BidPlaced"),
            JsonSubTypes.Type(value = BidPassed::class, name = "BidPassed"),
            JsonSubTypes.Type(value = BidWon::class, name = "BidWon"),
            JsonSubTypes.Type(value = LandedOnHostileSpace::class, name = "LandedOnHostileSpace"),
            JsonSubTypes.Type(value = RentDemanded::class, name = "RentDemanded"),
            JsonSubTypes.Type(value = RentPaid::class, name = "RentPaid"),
            JsonSubTypes.Type(value = TurnEnded::class, name = "TurnEnded"),
            JsonSubTypes.Type(value = OfferAdded::class, name = "OfferAdded"),
            JsonSubTypes.Type(value = OfferValueUpdated::class, name = "OfferValueUpdated"),
            JsonSubTypes.Type(value = OfferRemoved::class, name = "OfferRemoved"),
            JsonSubTypes.Type(value = TradeAccepted::class, name = "TradeAccepted"),
            JsonSubTypes.Type(value = TradeAcceptanceRevoked::class, name = "TradeAcceptanceRevoked"),
        ]
)
sealed class Event

data class GameCreated(
        val gameMaster: String,
        val board: List<Space>,
        val fixedStartMoney: Int,
        val interestRate: Double,
        val returnRate: Double
): Event()

data class PromotedToGameMaster(
    val player: String
): Event()

data class PlayerAdded(
        val id: String,
        val name: String,
        val color: String,
        val startDebt: Int
): Event()

data class GameStarted(
    val initiator: String
): Event()

data class NewTurnStarted(
        val player: String
): Event()

data class DiceRolled(
        val player: String,
        val dice1: Int,
        val dice2: Int
): Event()

data class StartMoneyReceived(
        val player: String,
        val amount: Int
): Event()

data class LandedOnSafeSpace(
    val player: String,
    val ground: String
): Event()

data class LandedOnBuyableSpace(
        val player: String,
        val ground: String
): Event()

data class SpaceBought(
    val ground: String,
    val player: String,
    val cash: Int,
    val borrowed: Int,
): Event()

data class BidStarted(
    val ground: String,
    val defaultWinner: String
): Event()

data class BidPlaced(
    val player: String,
    val bid: Int
): Event()

data class BidPassed(
    val player: String
): Event()

data class BidWon(
    val player: String,
    val bid: Int
): Event()

data class LandedOnHostileSpace(
        val player: String,
        val ground: String,
        val owner: String,
        val demandId: Int
): Event()

data class RentDemanded(
        val owner: String,
        val player: String,
        val rent: Int,
        val demandId: Int
): Event()

data class RentPaid(
    val player: String,
    val owner: String,
    val rent: Int,
    val demandId: Int
): Event()

data class TurnEnded(
    val player: String
): Event()

data class OfferAdded(
    val from: String,
    val to: String,
    val ownable: String,
    val value: Int
): Event()

data class OfferValueUpdated(
    val from: String,
    val to: String,
    val ownable: String,
    val value: Int
): Event()

data class OfferRemoved(
    val from: String,
    val to: String,
    val ownable: String
): Event()

data class TradeAccepted(
        val by: String,
        val with: String,
): Event()

data class TradeAcceptanceRevoked(
        val by: String,
        val with: String,
): Event()

File: frontend/src/app/game/api/event.ts
export type Event =
  GameCreated |
  PlayerAdded |
  PromotedToGameMaster |
  GameStarted |
  NewTurnStarted |
  DiceRolled |
  StartMoneyReceived |
  LandedOnSafeSpace |
  LandedOnBuyableSpace |
  SpaceBought |
  BidStarted |
  BidPlaced |
  BidPassed |
  BidWon |
  LandedOnHostileSpace |
  RentDemanded |
  RentPaid |
  TurnEnded |
  OfferAdded |
  OfferValueUpdated |
  OfferRemoved |
  TradeAccepted |
  TradeAcceptanceRevoked |
  TradeCompleted;

export interface GameCreated {
  type: 'GameCreated';
  gameMaster: string;
  board: Space[];
  fixedStartMoney: number;
  interestRate: number;
  returnRate: number;
}

export interface PlayerAdded {
  type: 'PlayerAdded';
  id: string;
  name: string;
  color: string;
  startDebt: number;
}

export interface PromotedToGameMaster {
  type: 'PromotedToGameMaster';
  player: string;
}

export interface GameStarted {
  type: 'GameStarted';
  initiator: string;
}

export interface NewTurnStarted {
  type: 'NewTurnStarted';
  player: string;
}

export interface DiceRolled {
  type: 'DiceRolled';
  player: string;
  dice1: number;
  dice2: number;
}

export interface StartMoneyReceived {
  type: 'StartMoneyReceived';
  player: string;
  amount: number;
}

export interface LandedOnSafeSpace {
  type: 'LandedOnSafeSpace';
  player: string;
  ground: string;
}

export interface LandedOnBuyableSpace {
  type: 'LandedOnBuyableSpace';
  player: string;
  ground: string;
}

export interface SpaceBought {
  type: 'SpaceBought';
  ground: string;
  player: string;
  cash: number;
  borrowed: number;
}

export interface BidStarted {
  type: 'BidStarted';
  ground: string;
  defaultWinner: string;
}

export interface BidPlaced {
  type: 'BidPlaced';
  player: string;
  bid: number;
}

export interface BidPassed {
  type: 'BidPassed';
  player: string;
}

export interface BidWon {
  type: 'BidWon';
  player: string;
  bid: number;
}

export interface LandedOnHostileSpace {
  type: 'LandedOnHostileSpace';
  player: string;
  ground: string;
  owner: string;
  demandId: number;
}

export interface RentDemanded {
  type: 'RentDemanded';
  owner: string;
  player: string;
  rent: number;
  demandId: number;
}

export interface RentPaid {
  type: 'RentPaid';
  player: string;
  owner: string;
  rent: number;
  demandId: number;
}

export interface TurnEnded {
  type: 'TurnEnded';
  player: string;
}

export interface OfferAdded {
  type: 'OfferAdded'
  from: string,
  to: string,
  ownable: string,
  value: number
}

export interface OfferValueUpdated {
  type: 'OfferValueUpdated'
  from: string,
  to: string,
  ownable: string,
  value: number
}

export interface OfferRemoved {
  type: 'OfferRemoved'
  from: string,
  to: string,
  ownable: string
}

export interface TradeAccepted {
    type: 'TradeAccepted';
    by: string
    other: string
}

export interface TradeAcceptanceRevoked {
  type: 'TradeAcceptanceRevoked' ;
  by: string
  other: string
}

export interface TradeCompleted {
  type: 'TradeCompleted'
  party1: string
  party2: string
  payments: Payment[]
  transfers: Transfer[]
}


export interface Payment {
  player: string,
  cashDelta: number,
  debtDelta: number
}

export interface Transfer {
  ownable: string
  from: string
  to: string
  value: number
}

type Space =
  Street |
  ActionSpace |
  Utility |
  Station |
  FreeParking |
  Prison;

interface Street {
  type: 'Street';
  id: string;
  text: string;
  color: string;
  initialPrice: number;
  rent: number;
  rentHouse: number[];
  rentHotel: number;
  priceHouse: number;
  priceHotel: number;
}

interface ActionSpace {
  type: 'ActionSpace';
  id: string;
  text: string;
}

interface Utility {
  type: 'Utility';
  id: string;
  text: string;
  initialPrice: number;
  rent: number;
  rentAll: number;
}

interface Station {
  type: 'Station';
  id: string;
  text: string;
  initialPrice: number;
  rent: number[];
}

interface Prison {
  type: 'Prison';
  id: string;
  text: string;
}

interface FreeParking {
  type: 'FreeParking';
  id: string;
  text: string;
}
